modules/go/main.go

```diff
@@ -34,57 +110,280 @@ type Go struct {
 
 	// Project source directory
 	Source *dagger.Directory
+
+	// Go module cache
+	ModuleCache *dagger.CacheVolume
+
+	// Go build cache
+	BuildCache *dagger.CacheVolume
+
+	// Base container from which to run all operations
+	Base *dagger.Container
+
+	// Pass arguments to 'go build -ldflags''
+	Ldflags []string
+
+	// Add string value definition of the form importpath.name=value
+	Values []string
+
+	// Enable CGO
+	Cgo bool
+
+	// Enable race detector
+	Race bool
 }
 
-// Build a base container with Go installed and configured
-func (p *Go) Base(
-	// Any extra apk packages that should be included in the base image
-	// +optional
-	extraPackages []string,
-) *dagger.Container {
-	pkgs := []string{
-		"go~" + p.Version,
-		// gcc is needed to run go test -race https://github.com/golang/go/issues/9918 (???)
-		"build-base",
-		// adding the git CLI to inject vcs info into the go binaries
-		"git",
-	}
-	pkgs = append(pkgs, extraPackages...)
-	return dag.
-		Wolfi().
-		Container(dagger.WolfiContainerOpts{Packages: pkgs}).
-		WithEnvVariable("GOLANG_VERSION", p.Version).
-		WithEnvVariable("GOPATH", "/go").
-		WithEnvVariable("PATH", "${GOPATH}/bin:${PATH}", dagger.ContainerWithEnvVariableOpts{Expand: true}).
-		WithDirectory("/usr/local/bin", dag.Directory()).
-		WithMountedCache("/go/pkg/mod", p.goModCache()).
-		WithMountedCache("/root/.cache/go-build", p.goBuildCache())
+// Download dependencies into the module cache
```

Comment:
Hm, this is odd - we shouldn't have a function that has side effects to populate a cache volume. These two systems interact in weird ways.

It's *totally* possible that this go mod download exec could cache, but the cache volume remains empty - that's confusing, but entirely expected.

Do we really need this function? If we really do, we should remove the cache volumes in this function before running it. I don't think it is needed though, it also feels like it exposes too much of the underlying go command to the user.

Reply:
Mm, maybe I'm misunderstanding the best practices. This `Download` is my interpretation of our current best practices, as a clean API.

The best practice in question: "download go dependencies as a separate command from building, to take advantage of cache volumes". You're right about the command caching - perhaps I'm missing a cache-buster?

Did I get the best practice wrong @sipsma @kpenfound ?

Reply:
I *think* this makes sense as an internal detail, for sure. Splitting the download and usage is good (although, a bit odd I suppose, since if the cache volume is actually empty and `go mod download` gets cached, then we end up downloading deps in the build step).

I think the issue is more that it shouldn't be a public API method for users to consume - it can just be a private method, since ideally this module should conceal how the sausage gets made.